performance/src/performance/details/is_same_object.cpp

```diff
@@ -0,0 +1,100 @@
+#include "traccc/performance/details/is_same_scalar.hpp"
+
+namespace traccc::details {
+
+/// @name Implementation for @c traccc::details::is_same_object<measurement>
+/// @{
+
+is_same_object<measurement>::is_same_object(const measurement& ref, scalar unc)
+    : m_ref(ref), m_unc(unc) {}
+
+bool is_same_object<measurement>::operator()(const measurement& obj) const {
+
+    return (is_same_scalar(obj.local[0], m_ref.get().local[0], m_unc) &&
+            is_same_scalar(obj.local[1], m_ref.get().local[1], m_unc) &&
+            is_same_scalar(obj.variance[0], m_ref.get().variance[0], m_unc) &&
+            is_same_scalar(obj.variance[1], m_ref.get().variance[1], m_unc));
+}
+
+/// @}
+
+/// @name Implementation for @c traccc::details::is_same_object<seed>
+/// @{
+
+is_same_object<seed>::is_same_object(
+    const spacepoint_container_types::const_view& ref_spacepoints,
+    const spacepoint_container_types::const_view& test_spacepoints,
+    const seed& ref, scalar unc)
+    : m_ref_spacepoints(ref.get_spacepoints(ref_spacepoints)),
+      m_spacepoints(test_spacepoints),
+      m_ref(ref),
+      m_unc(unc) {}
+
+bool is_same_object<seed>::operator()(const seed& obj) const {
+
+    // Extract the spacepoints belonging to the tested seed.
+    std::array<spacepoint, 3> test_spacepoints =
+        obj.get_spacepoints(m_spacepoints);
+
+    // Compare the two seeds.
+    return (is_same_scalar(obj.weight, m_ref.get().weight, m_unc) &&
+            is_same_scalar(obj.z_vertex, m_ref.get().z_vertex, m_unc) &&
+            is_same_object<spacepoint>(m_ref_spacepoints[0],
+                                       m_unc)(test_spacepoints[0]) &&
+            is_same_object<spacepoint>(m_ref_spacepoints[1],
+                                       m_unc)(test_spacepoints[1]) &&
+            is_same_object<spacepoint>(m_ref_spacepoints[2],
+                                       m_unc)(test_spacepoints[2]));
+}
+
+/// @}
+
+/// @name Implementation for @c traccc::details::is_same_object<spacepoint>
+/// @{
+
+is_same_object<spacepoint>::is_same_object(const spacepoint& ref, scalar unc)
+    : m_ref(ref), m_unc(unc) {}
+
+bool is_same_object<spacepoint>::operator()(const spacepoint& obj) const {
+
+    return (is_same_scalar(obj.x(), m_ref.get().x(), m_unc) &&
+            is_same_scalar(obj.y(), m_ref.get().y(), m_unc) &&
+            is_same_scalar(obj.z(), m_ref.get().z(), m_unc) &&
+            is_same_object<measurement>(m_ref.get().meas, m_unc)(obj.meas));
+}
+
+/// @}
+
+/// @name Implementation for
+///       @c traccc::details::is_same_object<bound_track_parameters>
+/// @{
+
+is_same_object<bound_track_parameters>::is_same_object(
+    const bound_track_parameters& ref, scalar unc)
+    : m_ref(ref), m_unc(unc) {}
+
+bool is_same_object<bound_track_parameters>::operator()(
+    const bound_track_parameters& obj) const {
+
+    return (is_same_scalar(obj.local()[0], m_ref.get().local()[0], m_unc) &&
```

Comment:
Can you also compare the output of `surface_link()` for bound track parameters

Reply:
Will do. :+1:

Reply:
This was eye-opening. As it turns out, it's exactly this comparison that made the `==` operator defined in Detray report none of the host and device objects matching each other.

You see, the surface link variable is not set anywhere in the traccc code right now. So far, so good. But while in the host algorithm this means that the default constructor would explicitly set the link to an invalid value, in device memory we don't call on the constructor to initialise any of the variables. So this variable was undefined in the output of the device algorithm. In practice being 0 with CUDA.

I set both the host and the device code to set the surface link to `0`. But now I'm tinking that that may have been a silly solution. Let me see if I can make it better on the device side...

Reply:
Right... there we go. This check made the device code quite a bit more robust. :wink:

Reply:
Oh I didn't notice that surface link was not assigned in track parameter estimation.
Assigning it to `geometry_id` (header of spacepoint) will do the job correctly.